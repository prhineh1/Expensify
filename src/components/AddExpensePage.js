import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense, this.props.uid)
        this.props.history.push('/');
    };
    render() {
        return (
        <div>
            <div className='page-header'>
                <div className='content-container'>
                    <h1 className='page-header__title'>Add Expense</h1>
                </div>
            </div>
            <div className='content-container'>
                <ExpenseForm
                onSubmit={this.onSubmit}
                budgets={this.props.budgets}
                />
             </div>
        </div>
        );
    }
}

const mapStateToProps = (state) => ({ 
    uid: state.auth.uid,
    budgets: state.budgets
});

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense, uid) => dispatch(addExpense(expense, uid))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);