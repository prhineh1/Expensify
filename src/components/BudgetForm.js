import React from 'react';

export default class BudgetForm extends React.Component {
  state = {
    description: this.props.budget ? this.props.budget.description : '',
    amount: this.props.budget ? (this.props.budget.amount / 100).toString() : '',
    error: ''
  };
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }))
  }
  onAmountChange = (e) => {
    const amount = e.target.value;
    if(!amount || amount.match(/^((([1-9])\d*)|0)(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }
  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please enter a description and amount.' }));
    }
    else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100
      });
    }
  }
  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          {this.state.error && <p>{this.state.error}</p>}
          <input
            type='text'
            placeholder='Budget Description'
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
           />
          <input
            type="text"
            placehoder='Budgeted Amount'
            value={this.state.amount}
            onChange={this.onAmountChange}
           />
        </form>
      </div>
    )
  }
}