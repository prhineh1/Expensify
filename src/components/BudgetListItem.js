import React from 'react';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import moment from 'moment';

const BudgetListItem = (props) => {
    const expensesTotal = props.expenses.reduce((sum, { amount }) => sum + amount, 0);
    const listItemClass = expensesTotal > props.budget.amount ? 'list-item list-item--over' : 'list-item';
    return ( 
        <Link className={listItemClass} to={'/edit/budget/'+ props.budget.id}>
            <div>
                <h3 className='list-item__title'>{props.budget.description}</h3>
            </div>
            <h4 className='list-item__budget'>
                <span className='list-item__budget__mobile'>Expenses: {numeral(expensesTotal / 100).format('$0,0.00')}</span>
                <span className='list-item__budget__desktop'>{numeral(expensesTotal / 100).format('$0,0.00')}</span>
            </h4>
            <h3 className='list-item__data'>
                <span className='list-item__data__mobile'>Budget: {numeral(props.budget.amount / 100).format('$0,0.00')}</span>
                <span className='list-item__data__desktop'>{numeral(props.budget.amount / 100).format('$0,0.00')}</span>
            </h3>
        </Link>
    )
};

export default BudgetListItem