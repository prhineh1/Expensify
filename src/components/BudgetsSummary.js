import React from 'react';
import { Link } from 'react-router-dom';

const BudgetsSummary = () => (
  <div className='page-header'>
      <div className='content-container'>
          <h1 className='page-header__title'>Budgets</h1>
          <p className='page-header__subtitle'>Note: Budgeted expenses are tracked monthly.</p>
          <div className='page-header__actions'>
              <Link className='button' to='/create/budget'>Add Budget</Link>
              <Link className='button' to='/dashboard/expenses'>View Expenses</Link>
          </div>
      </div>
  </div>
);

export default BudgetsSummary;