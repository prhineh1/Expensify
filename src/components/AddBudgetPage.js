import React from 'react';
import { connect } from 'react-redux';
import BudgetForm from './BudgetForm';
import { addBudget } from '../actions/budgets';

export class AddBudgetPage extends React.Component {
  onSubmit = (Budget) => {
    this.props.addBudget(Budget, this.props.uid);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
          <div className='page-header'>
              <div className='content-container'>
                  <h1 className='page-header__title'>Add Budget</h1>
                  <p className='page-header__subtitle'>Note: Budgeted expenses are tracked by month.</p>
              </div>
          </div>
          <div className='content-container'>
              <BudgetForm
              onSubmit={this.onSubmit}
              />
            </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({ uid: state.auth.uid });

export default connect(mapStateToProps, { addBudget })(AddBudgetPage);