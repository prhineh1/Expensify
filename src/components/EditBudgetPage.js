import React from 'react';
import { connect } from 'react-redux';
import { editBudget, removeBudget } from '../actions/budgets';
import BudgetForm from './BudgetForm';
import RemoveButtonModal from './RemoveButtonModal';

export class EditBudgetPage extends React.Component {
  state = {
      isOpen: false
  };
  onSubmit = (budget) => {
      this.props.editBudget(this.props.budget.id, budget, this.props.uid);
      this.props.history.push('/dashboard/budgets');
  };
  removeBudget = () => {
      this.props.removeBudget(this.props.budget.id, this.props.uid );
      this.props.history.push('/dashboard/budgets');
  };
  toggleModal = () => this.setState((prevState) => ({ isOpen: !prevState.isOpen }));
  render() {
      return (
      <div>
          <div className='page-header'>
              <div className='content-container'>
                  <h1 className='page-header__title'>Edit Budget</h1>
              </div>
          </div>
          <div className='content-container'>
              <BudgetForm
                  budget={this.props.budget}
                  onSubmit={this.onSubmit}
              />
              <button className='button button--secondary button--nav' onClick={this.toggleModal} >
                      Remove Budget
              </button>
          </div>
          <RemoveButtonModal 
              toggleModal={this.toggleModal} 
              removeBudget={this.removeBudget}
              isOpen={this.state.isOpen} 
          />
      </div>
      );
  }
};

const mapStateToProps = (state, props) => ({
  uid: state.auth.uid,
  budget: state.budgets.find(budget => budget.id === props.match.params.id)
});

export default connect(mapStateToProps, { editBudget, removeBudget })(EditBudgetPage);