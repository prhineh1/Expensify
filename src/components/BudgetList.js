import React from 'react';
import { connect } from 'react-redux';
import BudgetListItem from './BudgetListItem';
import moment from 'moment'; 
import partition from 'js-partition';

export class BudgetList extends React.Component {
    state = {
        displayIndex: 0
    };
    next = () => this.setState(prevstate => ({ displayIndex: prevstate.displayIndex + 1 }));
    prev = () => this.setState(prevstate => ({ displayIndex: prevstate.displayIndex - 1 }));
    render() {
        return(
            <div className='content-container'>
                <div className='list-header'>
                    <div className='show-for-mobile'>Budgets</div>
                    <div className='show-for-desktop'>Description</div>
                    <div className='show-for-desktop'>Expenses Total</div>
                    <div className='show-for-desktop'>Budgeted Amount</div>
                </div>
                <div className='list-body'>
                    {
                        this.props.budgets[this.state.displayIndex].length === 0 ? (
                            <div className='list-item list-item--message'>
                                <span>No budgets</span>
                            </div>
                        ) : (
                            this.props.budgets[this.state.displayIndex].map((budget) => 
                            <BudgetListItem 
                                key={budget.id} 
                                budget={budget} 
                                expenses={
                                    this.props.expenses.filter(expense => expense.budgetId === budget.id && 
                                    moment(expense.createdAt).isBetween(moment().startOf('month'), moment().endOf('month'), 'day'))
                                } 
                            />
                            )
                        )
                    }
                </div>
                {
                    !!this.props.budgets && (
                        <div className='list-nav'>
                            {
                                this.props.budgets[this.state.displayIndex + 1] && (
                                    <button className='button button--nav' onClick={this.next}>Next</button>
                                )
                            }
                            {  
                                this.props.budgets[this.state.displayIndex - 1] && (
                                    <button className='button button--nav' onClick={this.prev}>Previous</button>
                                )
                            }
                        </div>
                    )
                }
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    let partitionedBudgets;
    
    if (state.budgets.length % 10 === 0 && state.budgets.length > 0) {
        partitionedBudgets = partition(10, 10, state.budgets);
    }
    else {
        partitionedBudgets = partition(10, 10, [], state.budgets);        
    }
  return {
      expenses: state.expenses,
      budgets: partitionedBudgets
  };
};

export default connect(mapStateToProps)(BudgetList);