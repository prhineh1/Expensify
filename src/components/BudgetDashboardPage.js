import React from 'react';
import BudgetsSummary from './BudgetsSummary';
import BudgetList from './BudgetList';

const BudgetDashboardPage = () => (
  <div>
      <BudgetsSummary />
      <BudgetList />
  </div>
);

export default BudgetDashboardPage;