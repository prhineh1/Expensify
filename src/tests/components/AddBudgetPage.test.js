import React from 'react';
import { shallow } from 'enzyme';
import { AddBudgetPage } from '../../components/AddBudgetPage';
import budget from '../fixtures/budgets';

let wrapper, addBudget, history, uid;

beforeEach(() => {
  uid = 'abc123';
  addBudget = jest.fn();
  history =  { push: jest.fn() };
  wrapper = shallow(<AddBudgetPage addBudget={addBudget} history={history} uid={uid} />);
});

test('should render AddBudgetPage', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle addBudget', () => {
  wrapper.find('BudgetForm').prop('onSubmit')(budget);
  expect(history.push).toHaveBeenLastCalledWith('/dashboard/budgets');
  expect(addBudget).toHaveBeenLastCalledWith(budget, uid);
});