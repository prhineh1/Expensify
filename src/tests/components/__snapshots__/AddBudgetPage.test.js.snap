// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddBudgetPage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Budget
      </h1>
      <p
        className="page-header__subtitle"
      >
        Note: Budgeted expenses are tracked by month.
      </p>
    </div>
  </div>
  <div
    className="content-container"
  >
    <BudgetForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
